runtime/src/parachains.rs

```diff
@@ -179,14 +238,61 @@ decl_module! {
 
 			Self::check_attestations(&heads)?;
 
-			for head in heads {
+			for head in heads.iter() {
 				let id = head.parachain_index();
-				<Heads<T>>::insert(id, head.candidate.head_data.0);
+				<Heads<T>>::insert(id, &head.candidate.head_data.0);
 
 				// update egress.
 				for &(to, root) in &head.candidate.egress_queue_roots {
 					<Routing<T>>::insert((id, to), root);
 				}
+
+				// Queue up upwards messages (from parachains to relay chain).
+				if !head.candidate.upward_messages.is_empty() {
+					<RelayDispatchQueueSize<T>>::mutate(id, |&mut(ref mut count, ref mut len)| {
+						*count += head.candidate.upward_messages.len() as u32;
+						*len += head.candidate.upward_messages.iter()
+							.fold(0, |a, x| a + x.1.len()) as u32;
+					});
+					// Should never be able to fail assuming our state is uncorrupted, but best not
+					// to panic, even if it does.
+					let _ = <RelayDispatchQueue<T>>::append(id, &head.candidate.upward_messages[..]);
+				}
+			}
+
+			// simple round-robin dispatcher, using block number modulo parachain count
+			// to decide which takes precedence and proceeding from there.
+			let now = <system::Module<T>>::block_number();
+			let pids = Self::active_parachains();
+			let offset = (now % T::BlockNumber::from(Self::active_parachains().len() as u32))
```

Comment:
Consider the case when `Self::active_parachains().len() == 0`. It looks like the control can reach this point with this condition, can't it? If so, this will panic. If not, it would be cool to leave a comment here why (for future reviewers and auditors).